### ISSUE TITLE
Triangles are not set up correctly for segments

### ISSUE THREAD

**Issue by micha149** [TatumCreative/geo-3d-box#2]
I think every square needs to be compound of two triangles. But when I call `box = generateBox({ size: 5, segments: 2 })` I will get the following result:

![box-anim](https://cloud.githubusercontent.com/assets/298880/11117523/c81c09d0-893a-11e5-908a-7a0b1d4bc4c9.gif)

**Comment by gregtatum** [comment]
Ah darn, I'll take a look at it when I get a chance.

**Comment by gregtatum** [comment]
Are you using 1.0.4? This was a bug in an earlier version. I just tried it with those settings on my end and it worked just fine. Do you have source code I could look at?

<img width="870" alt="screen shot 2015-11-12 at 8 09 16 am" src="https://cloud.githubusercontent.com/assets/1588648/11120390/c92f5b9a-8914-11e5-825f-f04decca73bc.png">

**Comment by micha149** [comment]
Jep. `1.0.4` is installed. This is (hopefully) the interesting part of my code:

``` javascript
import createGeometry from 'gl-geometry';
import createShader from 'gl-shader';
import generateBox from 'geo-3d-box';
import normals from 'normals';
import mat4 from 'gl-mat4';

import Object3D from '../../object3d';
import fragShader from '../../shaders/bunny.frag';
import vertShader from '../../shaders/bunny.vert';
import Vector3 from '../../records/vector3';

class Box3D extends Object3D {
    constructor(gl) {
        super();
        this._gl = gl;
        this.size = 1;
        this.segments = 3;
        this.geometry = createGeometry(gl);
        this.shader = createShader(gl, vertShader, fragShader);
        this.isWireframe = false;
        this.x = 0;
        this.y = 0;
        this.z = 0;
    }

    update() {
        const box = generateBox({ size: this.size, segments: this.segments });

        this.geometry.attr('aPosition', box.positions);
        this.geometry.attr('aNormal', normals.vertexNormals(
            box.cells,
            box.positions
        ));
        this.geometry.faces(box.cells);
    }

    draw(camera) {
        const gl = this._gl;
        const model = mat4.create();
        const view = camera.view;
        const projection = camera.projection;

        mat4.translate(model, model, [this.x, this.y, this.z]);

        this.geometry.bind(this.shader);

        this.shader.uniforms.uProjection = projection;
        this.shader.uniforms.uView = view;
        this.shader.uniforms.uModel = model;

        this.geometry.draw(this.isWireframe ? gl.LINE_STRIP : gl.TRIANGLES);
    }
}
```

**Comment by gregtatum** [comment]
I think the issue (in your code) is the gl.LINE_STRIP is expecting the triangles in a different order than gl.TRIANGLES.

For an individual triangle the order needs to be modified like this: `[a,b,c] => [a,b,b,c,c,a]`

I have a utility using lodash to accomplish this in my own example:

```
var _ = require('lodash')

module.exports = function trianglesToLines( triangles ) {

    return _.chain( triangles )
        .chunk( 3 )
        .map( ([a,b,c]) => [ a,b,b,c,c,a ] )
        .flatten()
        .value()
}
```

**Comment by micha149** [comment]
Is this the correct way for drawing wireframes? I have this feature only for debugging reasons… Sorry for this question, I am totally new to webgl/gl/3d

**Comment by gregtatum** [comment]
Yeah the draw calls are order dependent for their elements. Check out https://msdn.microsoft.com/en-us/library/dn302396(v=vs.85).aspx these docs for the order things need to go in. You might have to reorder elements and manipulate the arrays to get the results you want depending on the drawing type you are doing.

**Comment by micha149** [comment]
Where do you get `triangles` for your box from? I suppose it's not `box.position`, or is it? The result with box.positions looks very strange…

**Comment by gregtatum** [comment]
The triangles are the indices of the vertices (positions) for the triangles, they get grouped in 3. Those are the box.cells.

**Comment by micha149** [comment]
Hey @TatumCreative. It's me again. Next question for the cells. You mentioned, they are grouped in 3. When I use `_.invert(box.cells, true)` I will get an object with all the groups. But there are some with 2 values and also with 6 values… I generated a box with size 2 and 2 segments. The value `0` can be found 2 times in the `cells` array. Value `4` can be found 6 times…

I tried http://stack.gl/packages/#hughsk/unindex-mesh to get an array representation of the box. This component also throws errors with my box mesh…
